33308: Eliminated AspectMock usage from ModuleResolverTest.php

anzin opened the pull request:
### Description

I've eliminated AspectMock usage from `dev/tests/unit/Magento/FunctionalTestFramework/Util/ModuleResolverTest.php`

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33308: [MFTF] Eliminate AspectMock from ModuleResolverTest (Complex!)

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

andrewbess reviewed (approved):
Hello @anzin 
Thank you for your job.

jilu1 commented:
@magento import pull request to https://github.com/magento-commerce/magento2-functional-testing-framework

magento-engcom-team commented:
@jilu1 the pull request successfully imported.

jilu1 commented:
@anzin 
Thank you for your PR! Please address unit test failure in ModuleResolverTest.

bohdan-harniuk reviewed (dismissed):
Hello, @anzin! 
Good job! Thank you for your contribution!

Hello, @jilu1!
Could you please check everything here? I've refactored the code a little bit. Now all tests passed successfully!

Thanks, Bohdan

jilu1 commented:
@magento import pull request to https://github.com/magento-commerce/magento2-functional-testing-framework

magento-engcom-team commented:
@jilu1 the pull request successfully imported.

jilu1 reviewed (changes requested):
@anzin @bohdan-harniuk 
Can we make sure to have the removed unit tests added back in for the new `ModuleResolverService`? 
For example, `testAggregateTestModulePathsDevTests`,  `testGetModulePathsLocations`, `testGetComposerJsonTestModulePathsForPathInvocation`, `testGetComposerInstalledTestModulePathsForPathInvocation `, etc

anzin commented:
Hello @jilu1 , sorry I don't understand your question. I removed these tests because we eliminate AspectMock, and these tests are not necessary, or I misunderstood something?

Thank you.

jilu1 commented:
> Hello @jilu1 , sorry I don't understand your question. I removed these tests because we eliminate AspectMock, and these tests are not necessary, or I misunderstood something?
> 
> Thank you.

@anzin Can you explain why these tests are not necessary?

anzin commented:
> ModuleResolverService

Hello, @jilu1!

Yes, I will try to add them!

Thank you!

anzin commented:
> > Hello @jilu1 , sorry I don't understand your question. I removed these tests because we eliminate AspectMock, and these tests are not necessary, or I misunderstood something?
> > Thank you.
> 
> @anzin Can you explain why these tests are not necessary? In other words, should you add these `testAggregateTestModulePathsDevTests`, `testGetModulePathsLocations`, `testGetComposerJsonTestModulePathsForPathInvocation`, `testGetComposerInstalledTestModulePathsForPathInvocation` as unit test for new class `ModuleResolverService`?

@jilu1, could you please help me to understand what does that test does? -> `testAggregateTestModulePathsDevTests`

I've investigated it again and came to the conclusion that this method tests nothing. Method `getModulesPath` run for the real instance of the `ModuleResolver` class but verified if the method `globRelevantPaths` was never called for the mocked version of this class. So, this test doesn’t make any sense. 
Could you please investigate and debug this method in the develop branch? I've removed these tests just because of the same reason.

Thanks!

anzin commented:
Hello, @jilu1, I've added new PR where returned deleted tests and fixed them so that they do not use AspectMock, please check if everything is well.

Thanks.

bohdan-harniuk reviewed (dismissed):
@anzin, thank you for your changes! For me it is fine!

@jilu1, please, please, check it!

Thanks, Bohdan!

jilu1 reviewed (commented):
@anzin 
Instead of calling methods directly from ModuleResolverService.php, you have kept several one-line wrapper methods in ModuleResolver.php after refactoring. Is this to help with unit testing?

anzin commented:
Hello @jilu1, my idea was to mock these methods because one test is did not passed, but then it turned out that it is not necessary to do so I put everything back.

Thanks.

jilu1 commented:
@anzin 

I understand why you moved some methods back. But my above question is actually why you keep wrapper methods like:
```
    private function getComposerInstalledTestModulePaths($composerFile)
    {
        return ModuleResolverService::getInstance()->getComposerInstalledTestModulePaths($composerFile);
    }

    private function getCustomModulePaths()
    {
        return ModuleResolverService::getInstance()->getCustomModulePaths();
    }
```
while you can call `ModuleResolverService::getInstance()->getComposerInstalledTestModulePaths($composerFile);` or `ModuleResolverService::getInstance()->getCustomModulePaths();` directly. Is it to help with unit test?

anzin commented:
@jilu1 , I understood what you mean, I've corrected the code please check if everything is well.

Thanks for your review.

andrewbess reviewed (approved):
Hello @anzin 
Thank you for your contribution.
Good job.

jilu1 commented on dev/tests/unit/Magento/FunctionalTestFramework/Util/ModuleResolverTest.php:
```diff
@@ -90,141 +103,130 @@ public function testGetModulePathsAggregate()
-                'Test' . DIRECTORY_SEPARATOR .'Mftf'
-            ]
-        );
+        $moduleResolverService = $this->createPartialMock(ModuleResolverService::class, ['globRelevantPaths']);
+        $moduleResolverService
+            ->method('globRelevantPaths')
+            ->withConsecutive()
+            ->will(
+                $this->returnCallback(
+                    function ($codePath, $pattern) use ($modulePath, $magentoBaseCodePath) {
+                        if ($codePath === $modulePath && $pattern === '') {
+                            return [];
+                        }
+
+                        if ($codePath === $magentoBaseCodePath . '/vendor' && $pattern === 'Test/Mftf') {
+                            return [];
+                        }
+
+                        if ($codePath === $magentoBaseCodePath . "/app/code" && $pattern === 'Test/Mftf') {
+                            return [];
+                        }
+
+                        $this->fail(sprintf(
+                            'Not expected parameter: \'%s\' when invoked method globRelevantPaths().',
+                            $modulePath
+                        ));
+                    }
+                )
+            );
```
@anzin 
This is close but does not match exactly what the original test does. The original test makes sure that all the paths parameters are used at least once, but the new test will miss that.

jilu1 reviewed (changes requested):
All look good except one comment. Please check.

anzin commented:
Hello @jilu1 , I've fixed test 'testGetModulePathsLocations', please check if everything is well.

Thanks.

jilu1 commented:
@magento import pull request to https://github.com/magento-commerce/magento2-functional-testing-framework

magento-engcom-team commented:
@jilu1 the pull request successfully imported.

bohdan-harniuk reviewed (dismissed):
@anzin, thank you for your work here!
